### PR TITLE
Refactor structure to enhance modularity of components

### PR DESCRIPTION
**Pull request by alexcu**
# Problem

It's really hard to find where things are as things aren't consistently named and the `templates` and `partials` folders are too big and messy. Changing one thing in one area may affect something somewhere else (especially CSS classes) and trying to find stuff can be a pain.

# What's changed

The entire web application directory structure is refactored with this PR (as well as some light code cleaning, and UI improvements :smile:)

By modularising code, the `templates` and `partials` folders for CoffeeScript and template HTML files are no longer separated—instead, the CoffeeScript, template HTML and LESS code for each component sits inside its own module folder as its own module, underneath the following root folders of:

- **`api`** - all API endpoints are separated into their own file
-**`common`** - all components that are re-used multiple times and do not have a strong association with a particular model
- **`config`** - all components related to configuration of the app (split from `app.coffee`)
- **`errors`** - all states for error handling
- **`groups`** - all directives and modals for handling group-related activities
- **`home`** - all states and modals related to the home page
- **`projects`** - all directives, states and modals for handling project-related activites
- **`sessions`** - all states and authentication factories for handling signing in/out
- **`tasks`** - all directives and modals for handling tasks-related activities, **including task/ILO alignment*
- **`units`** - all directives, states and modals for handling unit modification and batch unit listing activities such as student lists
- **`users`** - all modals and states for handling user-related activities
- **`visualisations`** - all directives responsible for visualising data (not the panels that showcase the data)

# Naming, coding and structure convention (try to follow so we don't run into this issue again)

For extendability and maintenance purposes, following these guidelines:

- Name a directive with it's role in mind (i.e., as a **[Agent Noun](https://en.wikipedia.org/wiki/Agent_noun)**) to give a small summary as to what the directive *does*:
  - when *viewing* a project or task, the directive is project-**viewer** and task-**viewer**
  - when *assessing* task submissions, the directive is task-submission-**assessor**
  - when *editing* a unit's tutorials, the directive is unit-tutorial-**editor**
- Name directives that show lots of data in one directive in a table a list: e.g.: `unit-student-list`, `group-member-list`, `project-top-task-list`
- Name directives with a series of steps to perform a goal a `wizard`, e.g.: `project-portfolio-wizard`, `new-user-wizard`
- Always name modals in Pascal Case `SomeModal` and create them as a factory/controller pair CoffeeScript file which can then be easily created on the fly:

```coffeescript
# foo/modals/create-foo-modal.coffee
angular.module('doubfire.foo.modals.create-foo-modal', [])

#
# Prompts the user to create a Foo using a bar and qux variable
#
.factory('CreateFooModal', ($modal) ->
  CreateFooModal = {}
  
  CreateFooModal.show = (bar, qux) ->
    $modal.open
      templateUrl: 'foo/modals/create-foo-modal.tpl.html'
      controller: 'CreateFooModalCtrl'
      resolve:
        bar: -> bar
        qux: -> qux

  CreateFooModal
)

.controller('CreateFooModalCtrl', ($scope, bar, qux) ->
  # Does stuff with bar and qux to create a foo
  $scope.bar = bar
  $scope.qux = qux
)
```

```coffeescript
# foo/states/foo-view/foo-view.coffee
# ...
.controller('FooViewCtrl', ($scope, CreateFooModal) ->
  # ...
  $createNewFoo = ->
    CreateFooModal.show($scope.bar, $scope.qux)
)
```

- Always name non-anonymous controllers with a `Ctrl` suffix
- Place modals and states in a `modals` and `states` folder under the root. All else can be in their own folders unless they are of a related concept (see the `project-portfolio-wizard` folder under `project`, `stats` under `tasks` and `units`) 
- The name of a module should follow the directory structure of where it has been placed (i.e., in the above example, the template file was at `foo/modals/create-foo-modal.tpl.html`, the CoffeeScript file was at `foo/modals/create-foo-modal.coffee`, and thus the module is `doubtfire.foo.modals.create-foo-modal`
- Try to give a brief summary of what the directive, state or factory does. E.g., the comment in the example above for `CreateFooModal` is sufficient.

---

Let me know what your suggestions are and if you see any room for improvement!

# TL;DR

Everything was a mess so it's been modularised and categorised. If we all follow these conventions from now on we will keep the codebase very clean and (hopefully) extendable and maintainable :smile:

### CODE REVIEW

**Comment by alexcu** [comment]
PS: @Reubsinit and @jakerenzella if you guys have a spare moment it'd be great if we can test everything to make sure nothing's broken.

**Comment by macite** [comment]
Looking good. I'll also test when I get a chance.

We need to capture all of the above in the details related to contributing to the project...
